relay-server/src/endpoints/common.rs

```diff
@@ -407,49 +413,89 @@ where
         .and_then(clone!(event_id, scoping, |project| {
             extract_envelope(&request, meta)
                 .into_future()
-                .and_then(clone!(project, |envelope| {
+                .and_then(clone!(event_manager, project_manager, |envelope| {
+                    // do dynamic sampling on transactions
                     event_id.replace(envelope.event_id());
 
+                    let trace_context = match envelope.get_trace_context() {
+                        None => {
+                            // if we don't have a trace context we can't do dynamic sampling so stop.
+                            return Box::new(Ok((envelope, None)).into_future())
+                                as ResponseFuture<(Envelope, Option<Addr<Project>>), _>;
+                        }
+                        Some(trace_context) => trace_context,
+                    };
+                    // Sample and potentially remove transactions from the envelope.
+                    // We only do this if the envelope contains only transactions,
+                    // The reason for that is that in case of envelopes containing only
+                    // transactions we have a chance to end processing here (if the transactions
+                    // are sampled out).
+                    // If the envelope contains other items (beyond transactions then we cannot
+                    // shortcut the processing here so we'll do it after we queue the envelope).
+                    let response = project_manager
+                        .send(GetProject {
+                            public_key: trace_context.public_key,
+                        })
+                        // deal with mailbox errors
+                        .map_err(BadStoreRequest::ScheduleFailed)
+                        // do the fast path transaction sampling (if we can't do it here
+                        // we'll try again after the envelope is queued)
+                        .and_then(clone!(event_manager, |project| {
+                            sample_transaction(envelope, Some(project.clone()), true)
+                                .map_err(clone!(event_id, |_| {
+                                    let event_id = event_id.borrow();
+                                    BadStoreRequest::TraceSampled(*event_id)
```

Comment:
Not quite sure what you mean ? 
For sample_transaction I only need to know the originating project for the trace context.

Reply:
That is true. But if you now issue a `BadStoreRequest::TraceSampled`, we jump into the `map_err` at the bottom which tries to emit an outcome. This outcome needs `scoping`, which would only be populated later.

I think we need to run `CheckEnvelope` first, and only then sampling.

Reply:
Yes you are right about the scoping (it might be only the scoping extracted from the meta, the partial scoping).

I have a problem calling CheckEnvelope before sampling because a CheckEnvelope enforces rate limits and might enforce rate limits for traces that are actually going to be sampled out. 

This would generate outcomes (for the transaction) that would be inaccurate (rate-limited rather then sampled).

**How about updating the scoping also inside sample_transaction(), i.e. calling Project::get_scoping() and updating the scoping info if the project state is available?**

Reply:
ok done